tests/unit/test-class-wc-payment-gateway-wcpay.php

```diff
@@ -1992,6 +1994,50 @@ public function test_process_payment_caches_mimimum_amount_and_displays_error_up
 		}
 	}
 
+	public function test_process_payment_rejects_if_missing_fraud_prevention_token() {
+		$order = WC_Helper_Order::create_order();
+
+		$fraud_prevention_service_mock = $this->getMockBuilder( Fraud_Prevention_Service::class )
+			->disableOriginalConstructor()
+			->getMock();
+		Fraud_Prevention_Service::set_instance( $fraud_prevention_service_mock );
+
+		$fraud_prevention_service_mock
+			->expects( $this->once() )
+			->method( 'is_enabled' )
+			->willReturn( true );
+
+		$this->expectException( Exception::class );
+		$this->expectExceptionMessage( 'Your payment was not processed.' );
+		$this->wcpay_gateway->process_payment( $order );
+	}
+
+	public function test_process_payment_rejects_if_invalid_fraud_prevention_token() {
+		$order = WC_Helper_Order::create_order();
+
+		$fraud_prevention_service_mock = $this->getMockBuilder( Fraud_Prevention_Service::class )
+			->disableOriginalConstructor()
+			->getMock();
+		Fraud_Prevention_Service::set_instance( $fraud_prevention_service_mock );
+
+		$fraud_prevention_service_mock
+			->expects( $this->once() )
+			->method( 'is_enabled' )
+			->willReturn( true );
+
+		$fraud_prevention_service_mock
+			->expects( $this->once() )
+			->method( 'verify_token' )
+			->with( 'incorrect-token' )
+			->willReturn( false );
+
+		$_POST['wcpay-fraud-prevention-token'] = 'incorrect-token';
+
+		$this->expectException( Exception::class );
+		$this->expectExceptionMessage( 'Your payment was not processed.' );
```

Comment:
Ideally, I'd like to assert exception code as well but it seems impossible. 

We are setting the [code when initially throwing the exception ](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payment-gateway-wcpay.php#L872)
```
throw new Process_Payment_Exception(
	__( 'Your payment was not processed.', 'woocommerce-payments' ),
	'rate_limiter_enabled'
);
```
And then in the `catch`	block we're[ re-throwing it but using only the message](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payment-gateway-wcpay.php#L952).
```
throw new Exception( WC_Payments_Utils::get_filtered_error_message( $e ) );
```
Is there a reason we shouldn't add the error code when re-throwing the exception?

Reply:
> Ideally, I'd like to assert exception code as well but it seems impossible.

Correct, that's due to the custom exception layer.

> Is there a reason we shouldn't add the error code when re-throwing the exception?

Looking into codebase and git history, no clues on my end.